File: Leetcode/234.palindromeLinkelist/234.cpp
 bool isPalindrome(ListNode* head) 
 {
 	if (!head || !head -> next) return true;
 	ListNode * slow, *fast;
 	slow = head;
 	fast = head;
 	while (fast -> next && fast -> next -> next)
 	{
 		fast = fast -> next -> next;
 		slow = slow -> next;
 	}
 	slow -> next = reverseList(slow -> next);
 	slow = slow -> next;
 	while(slow)
 	{
 		if (head -> val != slow -> val)
 			return false;
 		head = head->next;
 		slow = slow -> next;
 	}
 	return true;
 }

 	ListNode * reverseList (ListNode * head)
 	{
 		if (!head || !head -> next) return head;
 		ListNode * p = head -> next;
 		head ->next = NULL;
 		while (p)
 		{
 			ListNode * t = p -> next;
 			p -> next = head;
 			head = p;
 			p = t;
 		}
 		return head;
 	}

File: Leetcode/234.palindromeLinkelist/234.py
// bool False True;
class Solution(object):
     def reverseList(self,head):
    	if not head or not head.next :
    		return head
    	p = head.next
    	head.next = None# python None is NULL
    	while p :
    		t = p.next
    		p.next = head
    		head = p
    		p = t
        return head
        
     def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if not head or not head.next:
            return True
    	slow = fast = head
    	while fast.next and fast.next.next :
    		fast = fast.next.next
    		slow = slow.next
    	slow.next = self.reverseList(slow.next)//USE self.func to invoke the function
    	slow = slow.next
    	while slow :
    		if head.val != slow.val :
    			return False
    		head = head.next
    		slow = slow.next
        return True
